app/models/registration.rb

```diff
@@ -521,4 +529,62 @@ def competition_events_changed?
   def serializable_hash(options = nil)
     super(DEFAULT_SERIALIZE_OPTIONS.merge(options || {}))
   end
+
+  delegate :auto_accept_registrations, to: :competition
+
+  def auto_accept_in_current_env?
+    !(Rails.env.production? && EnvConfig.WCA_LIVE_SITE?)
+  end
+
+  def attempt_auto_accept
+    return false unless auto_accept_in_current_env?
+
+    failure_reason = auto_accept_failure_reason
+    if failure_reason.present?
+      log_auto_accept_failure(failure_reason)
+      return false
+    end
+
+    update_payload = build_auto_accept_payload
+    auto_accepted_registration = Registrations::RegistrationChecker.apply_payload(self, update_payload, clone: false)
+
+    if auto_accepted_registration.valid?
+      update_lanes!(
+        update_payload,
+        AUTO_ACCEPT_ENTITY_ID,
+      )
+      true
+    else
+      log_auto_accept_failure(auto_accepted_registration.errors.messages.values.flatten)
+      false
+    end
+  end
+
+  private def auto_accept_failure_reason
+    return Registrations::ErrorCodes::OUTSTANDING_FEES if outstanding_entry_fees.positive?
+    return Registrations::ErrorCodes::AUTO_ACCEPT_NOT_ENABLED unless competition.auto_accept_registrations?
+    return Registrations::ErrorCodes::INELIGIBLE_FOR_AUTO_ACCEPT unless competing_status_pending? || waiting_list_leader?
+    return Registrations::ErrorCodes::AUTO_ACCEPT_DISABLE_THRESHOLD if competition.auto_accept_threshold_reached?
+
+    Registrations::ErrorCodes::REGISTRATION_NOT_OPEN unless competition.registration_currently_open?
+  end
+
+  private def log_auto_accept_failure(reason)
+    add_history_entry(
+      { auto_accept_failure_reasons: reason },
+      'System',
+      AUTO_ACCEPT_ENTITY_ID,
+      'System reject',
+    )
+  end
+
+  private def build_auto_accept_payload
+    status = if competition.registration_full_and_accepted? && competing_status_pending?
+               Registrations::Helper::STATUS_WAITING_LIST
+             else
+               Registrations::Helper::STATUS_ACCEPTED
+             end
+
+    { user_id: user_id, competing: { status: status } }.with_indifferent_access
```

Comment:
Again, not happy about `with_indifferent_access`.

Reply:
Fixed! Had to use `deep_stringify_keys` as the hash is 2 levels deep

Reply:
Why do you need to have stringified keys here? Usually symbols are the more modern/ergonomic choice

Reply:
Also, if you really, _really_, _absolutely_ need string keys (you shouldn't), then you can directly create the hash with string keys: `{ "user_id" => user_id, "competing" => ... }`

Reply:
This is building a payload to be sent to `apply_payload` - which is built to accept `params`, which is string keys by default. So makes sense to me to keep the keys as strings

We could also handle this in apply_payload instead, but I think our naming/context makes it clear enough to expect to be dealing with string keys in that function?

Reply:
Ah, so that's kinda our fault. The `apply_payload` method mostly gets passed stuff from the controller's `params`, which are usually handed down with `indifferent_access` already baked in. So if anything, it's my fault for not refactoring it properly when I rebuilt `apply_payload` :P

Reply:
In that case, please at least go with instantiating string keys directly, instead of instantiating symbol keys and then direclty converting them to string keys on the spot

Reply:
It is now being instantiated with string keys :+1: